packages/swingset-theme-hashicorp/src/components/code-block/copy-button.tsx

```diff
@@ -0,0 +1,55 @@
+'use client'
+import {
+  IconDuplicate16,
+  IconFileCheck16,
+} from '@hashicorp/flight-icons/svg-react'
+import type { MouseEventHandler } from 'react'
+import { useState, useEffect, useCallback } from 'react'
+import { cx } from 'class-variance-authority'
+
+function CopyButton({ code }: { code: string }) {
+  const [isCopied, setIsCopied] = useState(false)
+
+  useEffect(() => {
+    const timeoutDuration = 2000
+    if (!isCopied) return
+    const timerId = setTimeout(() => {
+      setIsCopied(false)
+    }, timeoutDuration)
+
+    return () => {
+      clearTimeout(timerId)
+    }
+  }, [isCopied])
+
+  const handleClick = useCallback<
+    MouseEventHandler<HTMLButtonElement>
+  >(async () => {
+    setIsCopied(true)
+    if (!navigator?.clipboard) {
+      console.error('Access to clipboard rejected!')
+    }
+    try {
+      await navigator.clipboard.writeText(code)
+    } catch {
+      console.error('Failed to copy!')
+    }
+  }, [code])
```

Comment:
`useCallback` to cache function definition between renders. The `code` prop cannot change so once the function is defined, it's defined.